### PR TITLE
pivot table totals [WIP]

### PR DESCRIPTION
**Pull request by kdoh**
Addresses #5308 

Did a quick try at adding column and row totals to pivot tables. I'm sure there are some performance improvements we could make to what I did here.

screen shot 2017-06-09 at 6 03 26 pm

Questions:
- [ ] Should these be settings options?

Todos:

- [ ] make the total column and total footer sticky?
- [ ] add special styling to the total column and rows.
- [ ] tests

### CODE REVIEW

**Comment by mazameli** [comment]
This is a little exploration I'd done way back when we were doing the vertical query builder bar explorations. One additional thing it has is a "grand total" cell (just the sum of all the cells): 

pivot-table

File: frontend/src/metabase/lib/data_grid.js [hunk 1]
```diff
@@ -12,9 +12,9 @@ export function pivot(data) {
     // TODO: we assume dimensions are in the first 2 columns, which is less than ideal
     var pivotCol = 0,
         normalCol = 1,
-        cellCol = 2,
-        pivotColValues = distinctValues(data, pivotCol),
+        cellCol = 2, pivotColValues = distinctValues(data, pivotCol),
```
Comment (attekei) [inline frontend/src/metabase/lib/data_grid.js]: two lines accidentally combined together

File: frontend/src/metabase/lib/data_grid.js [hunk 2]
```diff
@@ -12,9 +12,9 @@ export function pivot(data) {
...
         normalColValues = distinctValues(data, normalCol);
+
     if (normalColValues.length <= pivotColValues.length) {
```
Comment (attekei) [inline frontend/src/metabase/lib/data_grid.js]: This function is long and has (unnecessary?) mutation which makes it hard to follow. For instance, I'm unable to decipher quickly what `normalColValues.length <= pivotColValues.length` condition means here.

Because this PR adds a remarkable amount of extra complexity, would be good to refactor parts of logic to smaller functions before a merge.

File: frontend/src/metabase/lib/data_grid.js [hunk 3]
```diff
@@ -37,12 +37,17 @@ export function pivot(data) {
...
     // make sure that the first element in the pivoted column list is null which makes room for the label of the other column
     pivotColValues.unshift(data.cols[normalCol].display_name);
 
+    // add a total column, this needs to happen here so that we make space for the eventual total
+    pivotColValues = pivotColValues.concat(["Total"])
+
     // start with an empty grid that we'll fill with the appropriate values
-    const pivotedRows = normalColValues.map((normalColValues, index) => {
+    let pivotedRows = normalColValues.map((normalColValues, index) => {
         const row = pivotColValues.map(() => null);
```
Comment (attekei) [inline frontend/src/metabase/lib/data_grid.js]: could this mapping method be in its own function?

Basically each `pivotedRows` transformation could be in its own mapping function, and the method call could be something like this:

```
const pivotedRows = normalColValues
    .map(createEmptyGrid)
    .map(populateWithData)
    .map(addTotalToRows)
    .map(addTotalToColumns)
```

This would conveniently also eliminate the mutation that makes reading the function confusing.

**Review by attekei — COMMENTED** [review]
I really like the feature & would maybe like to see a settings option for enabling/disabling it for either rows or columns. Also would be nice to have the "grand total" as suggested by Maz.

Left some refactoring suggestions for the `pivot` function!

**Comment by tanzoniteblack** [comment]
Will this only be addressing the use case for "total", or will we also be able to use this to do a couple different kind of aggregation functions (like `count distinct`)?

**Comment by tlrobinson** [comment]
I think we'd need to run two additional queries (with either of the two aggregations each) on the backend to be able to do most other aggregations besides sum/count.

**Comment by kstan79** [comment]
hi, was this merged? it is very useful hope can get it soon

**Comment by jbrooksuk** [comment]
Will we see this get resurrected at some point? This would be incredibly useful for us.
